ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveRelFieldTrimmer.java

```diff
@@ -315,48 +322,128 @@ private boolean isRexLiteral(final RexNode rexNode) {
+            Pair<RelOptTable, Integer> baseTable = Pair.of(tblRef.getTableRef().getTable(), tblRef.getTableRef().getEntityNumber());
+            if(candidateKeys.containsKey(baseTable)) {
+              List<Integer> candidateGBKeys = candidateKeys.get(baseTable);
+              candidateGBKeys.add(key);
+            } else {
+              List<Integer> candidateGBKeys =  new ArrayList<>();
+              candidateGBKeys.add(key);
+              candidateKeys.put(baseTable, candidateGBKeys);
+            }
+          }
+        }
       }
     }
-    if(currentKey == null || currentKey.isEmpty()) {
-      return originalGroupSet;
-    }
 
     // we want to delete all columns in original GB set except the key
     ImmutableBitSet.Builder builder = ImmutableBitSet.builder();
 
-    // we have established that this gb set contains keys and it is safe to remove rest of the columns
-    for(int i=0; i<backtrackedGBList.size(); i++) {
-      Integer backtrackedCol = backtrackedGBList.get(i);
-      int orgCol = originalGroupSet.nth(i);
-      if(fieldsUsed.get((orgCol))
-          || currentKey.get(backtrackedCol)) {
-        // keep the columns which are being used or are part of keys
-        builder.set(orgCol);
+    for(Map.Entry<Pair<RelOptTable, Integer>, Pair<List<Integer>, List<Integer>>> entry:mapGBKeysLineage.entrySet()) {
+      RelOptHiveTable tbl = (RelOptHiveTable)entry.getKey().left;
+      List<Integer> backtrackedGBList = entry.getValue().left;
+      List<Integer> gbKeys = entry.getValue().right;
+
+      ImmutableBitSet backtrackedGBSet = ImmutableBitSet.builder().addAll(backtrackedGBList).build();
+
+      List<ImmutableBitSet> allKeys = tbl.getNonNullableKeys();
+      ImmutableBitSet currentKey = null;
+      for(ImmutableBitSet key:allKeys) {
+        if(backtrackedGBSet.contains(key)) {
+          // only if grouping sets consist of keys
+          currentKey = key;
+          break;
+        }
+      }
+      if(currentKey == null || currentKey.isEmpty()) {
+        continue;
+      }
+
+      // we have established that this gb set contains keys and it is safe to remove rest of the columns
+      for(int i=0; i<backtrackedGBList.size(); i++) {
+        Integer backtrackedCol = backtrackedGBList.get(i);
+        int orgCol = gbKeys.get(i);
+        if(!fieldsUsed.get((orgCol))
+            && !currentKey.get(backtrackedCol)) {
+          // keep the columns which are being used or are part of keys
+          builder.set(orgCol);
+        }
+      }
+      // remove candidate keys if possible
+      if(candidateKeys.containsKey(entry.getKey())) {
+        List<Integer> candiateGbKeys = candidateKeys.get(entry.getKey());
+        for(Integer keyToRemove:candiateGbKeys) {
+          if(!fieldsUsed.get(keyToRemove)) {
+            builder.set(keyToRemove);
+          }
+        }
       }
     }
-    return builder.build();
+    ImmutableBitSet keysToRemove = builder.build();
+    return originalGroupSet.except(keysToRemove);
```

Comment:
Should we assert/check/throw error if the result is the empty set?

Reply:
Good idea.